src/main/java/com/pusher/client/channel/SubscriptionEventListener.java

```diff
@@ -39,7 +39,10 @@
      * @param data
      *            The JSON data that was included with the event. This can be
      *            parsed with Google's Gson library, which is a dependency of
-     *            this library, or your library of choice.
+     *            this library, or your library of choice
+     * @param event
+     *            A PusherEvent object which exposes the whole event.
+     *            See {@linkplain PusherEvent} for more.
      */
-    void onEvent(String channelName, String eventName, String data);
+    void onEvent(String channelName, String eventName, String data, PusherEvent event);
```

Comment:
What about just

```suggestion
    void onEvent(final PusherEvent event);
```

How does this compare to the swift/JS interface?

Reply:
I feel like it should be the way you suggest here, or passing channelName, eventName, data, and metadata or event data whatever you want to call it. The first time around we implemented it as the second suggestion I made here, but Tom/jim suggested the implementation we have now.

Reply:
Your the suggested API is nicer in the long-term IMO. However, leaving the existing params may help with migration; I think this is the main argument for leaving them?

Reply:
I also think @WillSewell's suggested API is more similar to the Swift API @TomKemp ?

Reply:
Yes, in the Swift library I've added a new method where you pass a callback that only receives a `PusherEvent`. I was considering making that the suggested way to bind an event and deprecating the existing method (but I'd still leave it in for backwards compatibility). 
`
// Existing
let _ = channel.bind(eventName: "client-test", callback: { (data: Any?) in
    //....
})
 
// New Addition
let _ = channel.bind(eventName: "client-test", eventCallback: { (event: PusherEvent) -> Void in
    //...
})
`

Reply:
For reference pusher-js does it in yet another way:

```
presenceChannel.bind('client-message', function (data, metadata) {
  console.log('received data from', metadata.user_id, ':', data);
});
```

Presumably for backwards compatibility.

---

Ideally we would have implemented the Java library in the same way as you did for swift @TomKemp, but due to the design of the library, it would mean users would have to implement both the new and legacy event handlers.

---

> However, leaving the existing params may help with migration

If we are making a breaking change anyway, I don't think breaking it in this way is going to make it particularly harder to migrate IMO.

The downside of leaving it as 

```
void onEvent(String channelName, String eventName, String data, PusherEvent event);
```

is it

* is less nice
* makes our websocket SDKs less consistent. It's hard to know how significant this is without having a principal we can apply.